src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
+            {
+                return false;
+            }
+
+            var iEquatableType = compilation.GetWellKnownType(WellKnownType.System_IEquatable_T);
+            if (iEquatableType is null)
+            {
+                return false;
+            }
+
+            var parameterType = method.Parameters[0].TypeWithAnnotations;
+            var constructedIEquatableType = iEquatableType.Construct(ImmutableArray.Create(parameterType));
+            var members = constructedIEquatableType.GetMembers(WellKnownMemberNames.ObjectEquals);
+            if (members.Length != 1)
+            {
+                // If the Equals method is missing on the interface definition, or it's overloaded, give up.
+                return false;
+            }
+
+            var constructedIEquatableEqualsMethod = members[0];
+            if (constructedIEquatableEqualsMethod.Equals(method))
+            {
+                return true;
+            }
+
+            var implementationMethod = method.ContainingType.FindImplementationForInterfaceMember(constructedIEquatableEqualsMethod);
+            return method.Equals(implementationMethod);
+        }
+
+        private bool IsIEqualityComparerEqualsMethodOrImplementation(MethodSymbol method)
+        {
+            if (method.ParameterCount != 2
+                || !method.Parameters[0].Type.Equals(method.Parameters[1].Type)
+                || method.MethodKind != MethodKind.Ordinary
+                || method.ReturnType.SpecialType != SpecialType.System_Boolean
+                || method.Name != WellKnownMemberNames.ObjectEquals)
+            {
+                return false;
+            }
+
+            var iEqualityComparerType = compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IEqualityComparer_T);
+            if (iEqualityComparerType is null)
+            {
+                return false;
+            }
+
+            var parameterType = method.Parameters[0].TypeWithAnnotations;
+            var constructedIEqualityComparerType = iEqualityComparerType.Construct(ImmutableArray.Create(parameterType));
+            var members = constructedIEqualityComparerType.GetMembers(WellKnownMemberNames.ObjectEquals);
+            if (members.Length != 1)
+            {
+                // If the Equals method is missing on the interface definition, or it's overloaded, give up.
+                return false;
+            }
+
+            var constructedIEqualityComparerEqualsMethod = members[0];
+            if (constructedIEqualityComparerEqualsMethod.Equals(method))
+            {
+                return true;
+            }
+
+            var implementationMethod = method.ContainingType.FindImplementationForInterfaceMember(constructedIEqualityComparerEqualsMethod);
+            return method.Equals(implementationMethod);
+        }
+
+        private void LearnFromEqualsMethod(BoundExpression left, TypeWithState leftType, BoundExpression right, TypeWithState rightType)
```

Comment:
>LearnFromEqualsMethod [](start = 21, length = 21)

Probably deserves a comment for documentation:
- a `null` argument tells us the other argument is maybe-null (ie. this is a pure null test)
- equality to an expression that is not-null means the argument is not-null

I see that the call to `LearnFromNullTest` in `AfterLeftChildHasBeenVisited` is preceded by a call to `SkipReferenceConversions`. I wonder if we need to do the same here. It would be good to test `IEquatable<B>.Equals(a, null)` where `a` has implicit reference conversion to type `B`. Testing with user-defined conversions which flip nullability would be useful too. 
We may have an issue tracking a problem in this area already.

Reply:
1. Should we expect `InstanceObjectEqualsMethod_MaybeNullExprReferenceConversion_NonNullExpr_NoWarningWhenTrue` to catch this scenario?
2. LearnFromNullTest also strips conversions internally (calls RemoveConversion on the expression ) so I'm also wondering if both steps are even necessary.

---
In reply to: [297388081](https://github.com/dotnet/roslyn/pull/36722#discussion_r297388081) [](ancestors = 297388081)

Reply:
Yes, `InstanceObjectEqualsMethod_MaybeNullExprReferenceConversion_NonNullExpr_NoWarningWhenTrue` is good.
I think we can probably remove `SkipReferenceConversions` from `AfterLeftChildHasBeenVisited` then. If a test is affected, then it can tell us what test to add for learning from Equals method.

---
In reply to: [297389419](https://github.com/dotnet/roslyn/pull/36722#discussion_r297389419) [](ancestors = 297389419,297388081)